units/host_status.go

```diff
@@ -117,45 +110,39 @@ clientsLoop:
 				j.AddError(errors.Wrap(err, "getting host statuses for providers"))
 				continue clientsLoop
 			}
-			if len(statuses) != len(hosts) {
-				j.AddError(errors.Errorf("programmer error: length of statuses != length of hosts"))
-				continue clientsLoop
-			}
+
 			for i := range hosts {
-				j.AddError(errors.Wrapf(j.setCloudHostStatus(ctx, m, hosts[i], statuses[i]), "setting status for host '%s' based on its cloud instance's status", hosts[i].Id))
+				hostID := hosts[i].Id
+				status, ok := statuses[hostID]
+				if !ok {
+					grip.Error(message.WrapError(err, message.Fields{
+						"message": "could not get any cloud instance status for host, defaulting to nonexistent status",
+						"host_id": hostID,
+						"job":     j.ID(),
+					}))
+					statuses[hostID] = cloud.StatusNonExistent
+				}
```

Comment:
This is a case where it's not supposed to happen (i.e. there's a problem with the `GetInstanceStatuses` implementation), but for resiliency against outages caused by `DescribeInstances` failing, it should still handle in some way. I opted to decide that if there's no status, that means the instance doesn't exist so that it will not keep trying to pick up the same instance (which has no status) over and over again.

I changed this to an alert instead of error so that we know if `GetInstanceStatuses` is violating the expected interface behavior.